File: lib/ts/src/lib.test.ts
import {test} from 'vitest';
import {ParentIPC} from './lib.js';

test('test connection timeout', async ({expect}) => {
    const parentIpc = new ParentIPC('../testdata/sleep15.sh', []);
    await expect(parentIpc.start()).rejects.toThrowError('timed out');
}, 15000);

test('test process stop before connection accept', async ({expect}) => {
    const parentIpc = new ParentIPC('../testdata/sleep3.sh', []);
    await expect(parentIpc.start()).rejects.toThrowError();
}, 15000);

File: lib/golang/lib_test.go
package golang

import (
	"os/exec"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
)

func TestNewParent(t *testing.T) {
	t.Run("socket argument in command", func(t *testing.T) {
		cmd := exec.Command("/bin/sh", ipcSocketArg, "/tmp/kek")
		_, err := NewParent(cmd)
		assert.Error(t, err)
	})

	t.Run("nonexistent binary", func(t *testing.T) {
		cmd := exec.Command("/nonexistent/binary")
		p, err := NewParent(cmd)
		assert.NoError(t, err)
		assert.Error(t, p.Start())
	})

	t.Run("connection timeout", func(t *testing.T) {
		cmd := exec.Command("../testdata/sleep15.sh")
		p, err := NewParent(cmd)
		assert.NoError(t, err)
		assert.Error(t, p.Start())
	})

	t.Run("child finished before accepting connection", func(t *testing.T) {
		cmd := exec.Command("../testdata/sleep3.sh")
		p, err := NewParent(cmd)
		assert.NoError(t, err)
		start := time.Now()
		assert.Error(t, p.Start())
		assert.WithinDuration(t, time.Now(), start, time.Second*4)
	})
}
